internal/session/repository_session.go

```diff
@@ -680,3 +674,54 @@ func fetchConnections(ctx context.Context, r db.Reader, sessionId string, opt ..
 	}
 	return connections, nil
 }
+
+// decryptAndMaybeUpdateSession switches between the database key and session key based on whether
+// the session uses a legacy private key or not. It also updates the encrypted session
+// in the database (if necessary). Eventually we should be able to remove this function
+// and use the session key unconditionally.
+func decryptAndMaybeUpdateSession(ctx context.Context, kmsRepo *kms.Kms, session *Session, writer db.Writer) error {
+	const op = "session.decryptAndMaybeUpdateSession"
+	sessionWrapper, err := kmsRepo.GetWrapper(ctx, session.ProjectId, kms.KeyPurposeSessions, kms.WithKeyId(session.KeyId))
+	if err != nil {
+		return errors.Wrap(ctx, err, op, errors.WithMsg("unable to get database wrapper"))
+	}
+	if len(session.CtCertificatePrivateKey) > 0 {
+		// New-style session with private key stored in DB, just decrypt.
+		if err := session.decrypt(ctx, sessionWrapper); err != nil {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("unable to decrypt session value"))
+		}
+		return nil
+	}
+	// No certificate private key present, this is a legacy session with a
+	// private key derived from the key. Derive it again and store it back
+	// in the DB.
+	_, session.CertificatePrivateKey, err = DeriveED25519Key(ctx, sessionWrapper, session.UserId, session.GetPublicId())
```

Comment:
Do we care if session.UserId or session.GetPublicId() are empty?  Derivation will succeed even if the salt and/or info are provided.

Reply:
Yep, I've added some validation that covers this.